Added Globus-Auth as an identity provider

NickolausDS opened the pull request.

NickolausDS committed:
Added Globus as an identity provider

* Globus Login work for any provider Globus Auth supports
* Added tests, coverage of new code is 100%
* Added Sphinx docs

NickolausDS commented on invenio_oauthclient/contrib/globus.py:
```diff
@@ -0,0 +1,226 @@
+    return user_info
+
+
+def get_user_id(remote, email):
```
Invenio binds user ids to usernames, so the Globus ID is also needed. This results in this second required call. The information _is_ stored within the id token, but one of the advantages to hitting the REST APIs directly in this case is we don't require any other dependencies to look inside the OIDC token or the Globus API.

ericblau replied:
Is it potentially problematic that Globus Auth explicitly states that the mapping between identity username and ID is not fixed over time?
https://docs.globus.org/api/auth/reference/#get_v2_api_identities_usernames_lt_list_of_identity_names_gt

ericblau replied:
Also, are we guaranteed to get the primary identity at index zero of the list from v2/api/identities?  it's not clear from the doc that order is guaranteed.  It may be possible for a Globus Auth user to have more than one identity that uses the same email address, in which case, I'm not sure that this get_user_id function necessarily returns what we want.

NickolausDS replied:
I didn't think it was possible for one identity username to map to multiple ids. The example shows calling in with multiple identity usernames and returning a bunch of corresponding ids to those usernames. 

For mapping identity usernames to IDs? That's a good question. If the identity username were to change hands, would the ID also change to something else thereby invalidating the user account on Invenio? Something I'm trying to confirm is that users in Invenio are tracked by external provider id rather than by username.

ericblau commented on examples/globus_app.py:
```diff
@@ -0,0 +1,155 @@
+        globus=globus.REMOTE_APP,
+    ),
+    GLOBUS_APP_CREDENTIALS=GLOBUS_APP_CREDENTIALS,
+    DEBUG=True,
```
Do we want DEBUG=True in general?  Or is this leftover from development?

NickolausDS replied:
I mostly followed the examples from all the other apps, which all set `DEBUG` to True.

NickolausDS commented on examples/globus_app.py:
```diff
@@ -0,0 +1,155 @@
+    SECURITY_PASSWORD_SALT='security-password-salt',
+    MAIL_SUPPRESS_SEND=True,
+    TESTING=True,
+    USERPROFILES_EXTEND_SECURITY_FORMS=True,
```
A couple notes on Config options. 

* Mail throws an error when a user logs in due to a bug in [Flask-Security](https://github.com/mattupstate/flask-security/issues/685). However, [Invenio is aware of the problem here](https://github.com/inveniosoftware/invenio-app-ils/issues/28). It's probably better to apply the fix in one of the dependent packages instead to avoid duplication. 
* MAIL_SUPPRESS_SEND also throws an error if it's set to `False`. This happens very rarely for logged in users and is probably better handled in downstream packages. I'm not completely sure why editing this value results in an error, but it seemed to be a [caching error here](https://github.com/mitsuhiko/flask-oauth/issues/48)
* SQLALCHEMY_TRACK_MODIFICATIONS throws a bunch of warnings, but should also probably be left to dependent packages to configure for themselves. We could set this in tests to squelch the warning messages on tests.

rpwagner commented on examples/globus_app.py:
```diff
@@ -0,0 +1,155 @@
+
+SPHINX-START
+
+1. Register a Globus application at `https://developers.globus.org/` with the
```
Add reference to the [app registration instructions](https://docs.globus.org/api/auth/developer-guide/#register-app)

NickolausDS replied:
There is a link on line 32, should it be more obvious?

rpwagner replied:
I think it's better to reference the app registration portion directly by adding the `#register-app` to the URL. More broadly, we should think about whether this is really just a developer activity. Deploying something like Invenio is a sys admin task.

NickolausDS replied:
Good point, I added links to those specific sections

rpwagner commented on invenio_oauthclient/contrib/globus.py:
```diff
@@ -0,0 +1,226 @@
+            consumer_secret='changeme',
+        )
+
+2. Register a Globus application at `https://developers.globus.org/` with the
```
Same here, reference the registration portion of the docs.

NickolausDS commented on tests/conftest.py:
```diff
@@ -88,6 +94,7 @@ def base_app(request):
         DEBUG=False,
         SECRET_KEY='TEST',
         SECURITY_DEPRECATED_PASSWORD_SCHEMES=[],
+        SQLALCHEMY_TRACK_MODIFICATIONS=False,
```
I should note that this isn't part of Globus Auth, but tracking modifications in tests isn't needed and setting this variable prevents a large number of warnings produced every time a test is run.

ericblau commented:
Looks good to me, now.

NickolausDS commented:
I went ahead and rebased on the latest upstream changes and squashed the commits into one, along with a few descriptive bullet points. This should be good to go.

NickolausDS commented:
@ericblau @rpwagner @lukaszlacinski Closing this PR and opening up the official one. Thanks for all your feedback!
